3.0.x php unit testing fixes pre rfc

simon-mundy opened the pull request:
Upgraded tests using rector -> PHPUnit 11
Fixed datasources to ensure they are static
Various refactoring to tests for compliance
Made E_NOTICE/E_ERROR silent for testing purposes
Commented out incomplete tests

For all intents and purposes this means the unit testing should work as-is. There are still numerous upgrades to perform - namely a lot of deprecated getMockForAbstractClass() calls in the tests - but they will need to wait until the satellite packages have been moved out as it's a lot of work.

simon-mundy committed:
Upgraded tests using rector -> PHPUnit 11
Fixed datasources to ensure they are static
Various refactoring to tests for compliance
Made E_NOTICE/E_ERROR silent for testing purposes

simon-mundy committed:
Upgraded tests using rector -> PHPUnit 11
Fixed datasources to ensure they are static
Various refactoring to tests for compliance
Made E_NOTICE/E_ERROR silent for testing purposes
Commented out incomplete tests

simon-mundy committed:
Upgraded tests using rector -> PHPUnit 11
Fixed datasources to ensure they are static
Various refactoring to tests for compliance
Made E_NOTICE/E_ERROR silent for testing purposes
Commented out incomplete tests

simon-mundy committed:
Upgraded tests using rector -> PHPUnit 11
Fixed datasources to ensure they are static
Various refactoring to tests for compliance
Made E_NOTICE/E_ERROR silent for testing purposes
Commented out incomplete tests

simon-mundy committed:
Merged

simon-mundy committed:
Upgraded tests using rector -> PHPUnit 11
Fixed datasources to ensure they are static
Various refactoring to tests for compliance
Made E_NOTICE/E_ERROR silent for testing purposes
Commented out incomplete tests

tyrsson commented on test/integration/Adapter/Driver/Pdo/AbstractAdapterTestCase.php:
```diff
@@ -11,13 +12,11 @@
-     */
     public function testConnection()
     {
         $this->assertInstanceOf(Adapter::class, $this->adapter);
```
When possible we really need to start moving this usage to AdapterInterface. Of course it might make sense to address typing changes in focused PR's where those changes can be isolated based on Interface.

tyrsson replied:
I'm marking this as resolved since the more I consider it the more I think the typing changes should be isolated in focused PR's.

tyrsson commented on test/integration/Adapter/Driver/Pdo/Mysql/AdapterTrait.php:
```diff
@@ -26,8 +25,7 @@ protected function setUp(): void
 
-    /** @return null|string */
-    protected function getHostname()
+    protected function getHostname(): ?string
```
This needs expanded to include all possible returns types of getenv which iirc is array|string|false

simon-mundy replied:
Done

tyrsson commented on test/integration/Adapter/Driver/Pdo/Mysql/TableGatewayTest.php:
```diff
@@ -5,26 +5,26 @@
-    /**
-     * @covers \Laminas\Db\TableGateway\TableGateway::__construct
-     */
     public function testConstructor()
```
These need return types of void if another test is not dependent on their return.

tyrsson commented on test/integration/Adapter/Driver/Pdo/Postgresql/AdapterTrait.php:
Instead of the current usage on line 13/14 I think we might be better served by using

https://docs.phpunit.de/en/11.5/attributes.html#requiresphpextension

simon-mundy replied:
Yes we will but it will need to be better handled on a per-satellite-package basis.

tyrsson commented on test/integration/Adapter/Platform/MysqlTest.php:
In my previous comment I mentioned the requires extension attribute which if cleverly used would remove the need for all of the boilerplate in this class.

tyrsson commented on test/integration/Platform/PgsqlFixtureLoader.php:
```diff
@@ -12,16 +12,16 @@
 
-    /** @var PDO */
-    private $pdo;
+    private PDO $pdo;
```
Lets be consistent with these. Either use statements for all or for none. The also need to accept null since in the disconnect method of all of the Loaders they are assigned null.

simon-mundy replied:
I'm not sure what you mean about this?

tyrsson replied:
In some places were using \PDO vs PDO.

In regards to the it needs to accept null.

private ?PDO $pdo since in the disconnect method of that class we have:

$this->pdo = null;

tyrsson commented on test/integration/Platform/SqlServerFixtureLoader.php:
These and other classes will need to be marked as final where appropriate.

simon-mundy replied:
Done

tyrsson commented on test/unit/Adapter/AdapterAbstractServiceFactoryTest.php:
```diff
@@ -8,12 +8,16 @@
 {
     /** @var ServiceLocatorInterface */
-    private $serviceManager;
+    private ServiceManager|ServiceLocatorInterface $serviceManager;
```
We should be typing to Psr\Container\ContainerInterface here. Mezzio supports many containers other than ServiceManager

simon-mundy replied:
I'm going to mark this as a @TODO as doing this breaks current tests

tyrsson commented on test/unit/Adapter/AdapterAwareTraitTest.php:
We need coverage information on the traits

tyrsson commented on test/unit/Adapter/Driver/IbmDb2/ConnectionIntegrationTest.php:
```diff
@@ -5,28 +5,33 @@
+use PHPUnit\Framework\Attributes\CoversMethod;
+use PHPUnit\Framework\Attributes\Group;
 
 use function ini_get;
```
Attributes\RequiresSetting

tyrsson commented on test/unit/Adapter/Driver/IbmDb2/IbmDb2Test.php:
```diff
@@ -99,10 +106,9 @@ public function testGetConnection()
      * @todo   Implement testGetPrepareType().
      */
-    public function testCreateStatement()
+    public function testCreateStatement(): never
```
I have to be overlooking something here but I currently do not understand how this does not cause an error by using a never return type.

settermjd replied:
I think it makes sense, @tyrsson, based on my reading of [the docs](https://www.php.net/manual/en/language.types.never.php).

tyrsson replied:
Now, after doing some digging I see why it works. The markTestIncomplete method (or at least the one I could find, which is static and a whole different discussion), is return typed as never due to its only purpose is throwing a IncompleteTestError. I guess it pays to know more about phpunit internals.

_goes to see how much grief, if any, this will cause with psalm and no baseline_

tyrsson commented on test/unit/Adapter/Driver/IbmDb2/AbstractIntegrationTestCase.php:
These integration test need moved.

tyrsson commented on test/unit/Adapter/Driver/IbmDb2/StatementTest.php:
```diff
@@ -6,18 +6,26 @@
 
 use function error_reporting;
 
 include __DIR__ . '/TestAsset/Db2Functions.php';
```
I would suggest split these into their own files and autoload the files in the autoload-dev block.

tyrsson commented on test/unit/Adapter/Driver/Oci8/ConnectionIntegrationTest.php:
```diff
@@ -141,22 +130,20 @@ public function testExecute()
-     * @group laminas3469
-     */
-    public function testConnectReturnsConnectionWhenResourceSet()
+    #[Group('laminas3469')]
```
Is this group really needed?

settermjd replied:
What does  the number stand for?

tyrsson commented on test/unit/Adapter/Driver/Pdo/ConnectionTest.php:
```diff
@@ -34,25 +35,21 @@ public function testResource()
-    /**
-     * @group 2622
-     */
+    #[Group('2622')]
```
I'm unsure of what benefit these random groups are supposed to be. Have you found any purpose for these?

settermjd replied:
Perhaps they meant something at one point. However, given the lack of anything that stands out, anything intuitive, as you say @tyrsson is there any sense in keeping them?

tyrsson replied:
It seems as if at one time they may have referenced issues, PR's or something to that effect. I can't seem to find reference to most of the group numbers at this point though so I'm not sure what purpose they can serve. @simon-mundy may remember why they were used.

I can see a benefit for grouping test but only if we know why that grouping is relevant.

simon-mundy replied:
I can only assume they were related to issue numbers?

tyrsson commented on test/unit/Adapter/Platform/IbmDb2Test.php:
```diff
@@ -123,24 +111,21 @@ public function testQuoteTrustedValue()
-    /**
-     * @covers \Laminas\Db\Adapter\Platform\IbmDb2::quoteValueList
-     */
     public function testQuoteValueList()
```
Attributes\RequiresFunction

simon-mundy replied:
Fixed

tyrsson commented on test/unit/Adapter/Platform/OracleTest.php:
```diff
@@ -131,30 +124,25 @@ public function testQuoteIdentifierChain()
-            . 'extension/driver support can introduce security vulnerabilities in a production environment'
-        );
+        /**
+         * @todo Determine if vulnerability warning is required during unit testing
```
On these I would say we at least need to cover that under the right conditions a notice is provided.

simon-mundy replied:
I'm not quite sure how to achieve this - PHPUnit has phased out any of the WARNING|NOTICE|DEPRECATED methods now.

tyrsson commented on test/unit/Adapter/Platform/SqliteTest.php:
```diff
@@ -116,30 +106,24 @@ public function testQuoteTrustedValue()
+        /**
+         * @todo Determine if vulnerability warning is required during unit testing
+         */
+        //$this->expectError();
```
Why would we expect an Error here when others have expected a notice?

simon-mundy replied:
I don't know without more rigorous testing, but it may not even be required once we pull these into satellites.

tyrsson commented on test/unit/ConfigProviderTest.php:
```diff
@@ -4,13 +4,14 @@
 {
     /** @var array<string, array<array-key, string>> */
-    private $config = [
+    private array $config = [
```
Should we remove all references to Zend\* packages? Zend bridge has been removed from all packages as a dep for awhile IIRC.

simon-mundy replied:
Yep. Done.

tyrsson commented on test/unit/Sql/Platform/Mysql/SelectDecoratorTest.php:
```diff
@@ -15,32 +15,32 @@
-     * @dataProvider dataProvider
-     */
+    protected Sql $sql;
+    #[DataProvider('dataProvider')]
```
Need a blank line here.

tyrsson commented on test/unit/Sql/SqlTest.php:
```diff
@@ -208,33 +194,21 @@ public function testForDifferentAdapters()
-                $platform = null;
-        }
+        $platform = match ($platform) {
+            'sql92' => new TestAsset\TrustingSql92Platform(),
```
The => probably need to be aligned here for coding standard.

simon-mundy replied:
Weirdly running PHPCBF did not identify this as needing fixing! Have manually done so.

tyrsson commented on test/unit/Sql/TableIdentifierTest.php:
```diff
@@ -40,10 +42,7 @@ public function testGetSchema()
 
     public function testGetTableFromObjectStringCast()
```
I'm unsure what the point of this entire test is.

tyrsson commented on test/unit/Sql/TableIdentifierTest.php:
```diff
@@ -52,32 +51,23 @@ public function testGetTableFromObjectStringCast()
 
     public function testGetSchemaFromObjectStringCast()
```
Again, I'm unsure what the point of these test are. Seems as they are dependent on the object to be cast not the TableIdentifier accepting them. The TableIdentifier only accepts strings so it seems as if these are just writing test to be writing test I see no benefit in coverage from them as they do not test any behavior of the SUT.

simon-mundy replied:
Agree, it seems contrived for no good reason, I only solved this to squish errors.

tyrsson commented on test/unit/TestAsset/ObjectToString.php:
I'm not so sure we need this if the only test depending on it is the two test for TableIdentifier.

tyrsson reviewed (changes requested):
Overall looks very good. (I know you were just fixing what was already there)

For all of the test cases they need renamed to SomeTestCase and if test do not extend them they need to be marked as final.

All test methods that are not returning values need typed as :void. I'm requesting this change here since we are already touching all of these files and it will improve the psalm baseline a lot.

I would also like to get @settermjd thoughts on this review. It's a lot to review so maybe he can just provide some feedback on this one.

@simon-mundy Thank you for the huge amount of work in this PR.

tyrsson commented on test/integration/Platform/MysqlFixtureLoader.php:
```diff
@@ -12,13 +12,14 @@
 
-    /** @var PDO */
-    private $pdo;
+    private PDO $pdo;
```
These, in all of the *FixtureLoader classes must be nullable.

tyrsson replied:
I updated my other comment related to these so there is no sense having it spread into two comments. Marking this resolved.

settermjd commented on test/integration/Adapter/Driver/Pdo/Mysql/TableGatewayTest.php:
```diff
@@ -66,7 +62,7 @@ public function testInsert()
      * @see https://github.com/zendframework/zend-db/pull/178
      *
-     * @return mixed
+     * @return int
```
Why not set this as the function's return type?

simon-mundy replied:
Done

settermjd commented on test/unit/Adapter/AdapterServiceFactoryTest.php:
```diff
@@ -5,19 +5,21 @@
 {
-    /** @var ServiceLocatorInterface&MockObject */
-    private $services;
+    private ServiceLocatorInterface&MockObject $services;
```
Does this need to still be a `ServiceLocatorInterface` object, or can it be a `ContainerInterface`?

simon-mundy replied:
Won't fix (likely that this is made redundant anyway)

settermjd commented on test/unit/Adapter/AdapterTest.php:
```diff
@@ -175,54 +176,46 @@ public function testCreatePlatform()
+     * @throws Exception
+     * @throws \Exception
+     */
+    #[Group('#210')]
```
What does 210 stand for?

simon-mundy replied:
Assuming it's to do with tests from old bug reports?

settermjd commented on test/unit/Adapter/Driver/IbmDb2/AbstractIntegrationTestCase.php:
```diff
@@ -7,10 +7,10 @@
 use function getenv;
 
-abstract class AbstractIntegrationTest extends TestCase
+abstract class AbstractIntegrationTestCase extends TestCase
```
👍🏽

settermjd commented on test/unit/Adapter/Driver/IbmDb2/TestAsset/Db2Functions.php:
```diff
@@ -47,13 +39,9 @@ function db2_stmt_errormsg($stmt = null)
- * @return string
  */
-function db2_stmt_error($stmt = null)
+function db2_stmt_error(mixed $stmt = null): string
```
Is it possible to be more specific, instead of using `mixed`?

simon-mundy replied:
Won't fix for now, this will be a satellite package issue.

settermjd commented on test/unit/Adapter/Driver/Mysqli/ConnectionTest.php:
```diff
@@ -132,9 +126,8 @@ public function testConnectionFails()
-     * @return MockObject
      */
-    protected function createMockMysqli($flags)
+    protected function createMockMysqli(int $flags): MockObject
```
Wouldn't this return a union type?

simon-mundy replied:
This test is likely to be rewritten and currently passes tests as-is.

settermjd commented on test/unit/Adapter/Driver/Pdo/ConnectionIntegrationTest.php:
```diff
@@ -152,19 +137,14 @@ public function testPrepare()
         $this->markTestIncomplete('Need to create a temporary sequence.');
-        $connection = new Connection($this->variables);
-        $connection->getLastGeneratedValue();
+        //$connection = new Connection($this->variables);
+        //$connection->getLastGeneratedValue();
```
Why keep them commented out?

simon-mundy replied:
PHPStorm complains about code that never gets executed (bc of the `markTestIncomplete` method returns a never). I didn't want to lose anything that was there, hence commenting it out for now. It's a temporary measure.

settermjd commented on test/unit/Adapter/Driver/Pdo/ConnectionTest.php:
```diff
@@ -64,7 +61,7 @@ public function testArrayOfConnectionParametersCreatesCorrectDsn()
         try {
             $this->connection->connect();
-        } catch (Exception $e) {
+        } catch (Exception) {
```
I think I'm forgetting something, but what's the sense in catching an exception only to do nothing?

tyrsson replied:
I think because of the way this test is structured we are not interested in whether it throws an exception. If the exception is not caught then our test would end and our assertions would never take place for the constructed dsn.

simon-mundy replied:
Also this entire test suite is likely to be rewritten anyway.

settermjd reviewed (changes requested):
Just a few comments, mate Otherwise, thanks for the hard work.

simon-mundy committed:
Updates to PR#38

simon-mundy committed:
Updates to PR#38

simon-mundy committed:
Updates to PR#38

simon-mundy committed:
Updates to PR#38

simon-mundy committed:
Updates to PR#38

simon-mundy committed:
Updates to PR#38

simon-mundy committed:
Updates to PR#38

simon-mundy committed:
Updates to PR#38

simon-mundy committed:
Updates to PR#38

tyrsson commented on src/Sql/Delete.php:
```diff
@@ -32,8 +32,8 @@ class Delete extends AbstractPreparableSql
-    /** @var string|TableIdentifier */
-    protected $table = '';
+    /** @var string|array|TableIdentifier */
+    protected TableIdentifier|string|array $table = '';
```
How can this accept an array?

simon-mundy committed:
Upgraded tests using rector -> PHPUnit 11
Fixed datasources to ensure they are static
Various refactoring to tests for compliance
Made E_NOTICE/E_ERROR silent for testing purposes
Commented out incomplete tests

simon-mundy committed:
Upgraded tests using rector -> PHPUnit 11
Fixed datasources to ensure they are static
Various refactoring to tests for compliance
Made E_NOTICE/E_ERROR silent for testing purposes
Commented out incomplete tests

simon-mundy committed:
Upgraded tests using rector -> PHPUnit 11
Fixed datasources to ensure they are static
Various refactoring to tests for compliance
Made E_NOTICE/E_ERROR silent for testing purposes
Commented out incomplete tests

simon-mundy committed:
Upgraded tests using rector -> PHPUnit 11
Fixed datasources to ensure they are static
Various refactoring to tests for compliance
Made E_NOTICE/E_ERROR silent for testing purposes
Commented out incomplete tests
